errors.go

```diff
@@ -47,9 +55,34 @@ const (
 	ErrUnknown            = "unknown"
 )
 
-// Error returns a string message of the error. It is a concatenation of Code and Message params
-// This means the Error implements the error interface
+// Error returns a string message of the error.
+// It will contain the code and error message. If there is a causal chain, the
+// message from each error in the chain will be added to the output.
 func (p *Error) Error() string {
+	if p.cause == nil {
+		// Not sure if the empty code/message cases actually happen, but to be safe, defer to
+		// the 'old' error message if there is no cause present (i.e. we're not using
+		// new wrapping functionality)
+		return p.legacyErrString()
+	}
+	var next error = p
+	output := strings.Builder{}
+	output.WriteString(p.Code)
+	for next != nil {
+		output.WriteString(": ")
+		switch typed := next.(type) {
+		case *Error:
+			output.WriteString(typed.Message)
```

Comment:
We don't also print the code of the causes?

Reply:
Good question! The code of the cause will most likely be the same all the way down the chain until the bottom, where it probably is different. If you print them all you just get something like:

```
internal_service: foo: internal_service: bar ...
```

Which is quite hard to read. I wonder if it would make sense to print the code of the _original_ cause if it was a terror through. Or potentially change the format of the printing completely so that it would be more readable.